build_tools/cmake/iree_test_matrix.cmake

```diff
@@ -0,0 +1,73 @@
+# Copyright 2021 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+include(CMakeParseArguments)
+
+# Generates a matrix of permuted test cases from a desription in the
+# test_matrix.yaml file in the current source directory.
+# The resulting test files will be written to:
+#   ${CMAKE_CURRENT_BINARY_DIR}/generated
+# Typically this should be followed by:
+#   iree_test_matrix_glob_py_tests
+#
+# (type of tests to add are based on the runners defined in the
+# generation script)
+function(iree_test_matrix_gen)
+  set(_test_matrix_yaml "${CMAKE_CURRENT_SOURCE_DIR}/test_matrix.yaml")
+  set(_generator_script "${PROJECT_SOURCE_DIR}/build_tools/testing/gen_test_matrix.py")
+  message(STATUS "Generating tests for ${CMAKE_CURRENT_SOURCE_DIR}")
+  execute_process(
+    COMMAND "${Python3_EXECUTABLE}"
+      "${_generator_script}"
+      --dir "${CMAKE_CURRENT_SOURCE_DIR}"
+      --output_dir "${CMAKE_CURRENT_BINARY_DIR}"
+    WORKING_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}"
+    RESULTS_VARIABLE _gen_result)
```

Comment:
Ah, yeah. I can make that work, but I'll do it in a follow-up where I can try it a few times and make sure it is robust:

```
# TODO: Add a `generate` step conditioned on the test_matrix.yaml file, so that
# reconfigure is automatic.
```